proxy/src/project.rs

```diff
@@ -80,14 +86,138 @@ pub enum Registration {
+    fn to_helper(&self) -> String {
+        match self {
+            Self::Password(pass) => format!(
+                "credential.helper=!f() {{ test \"$1\" = get && echo \"password={}\"; }}; f",
+                pass
+            ),
+        }
+    }
+}
+
+/// The data necessary for checking out a project.
+pub struct Checkout<P>
+where
+    P: AsRef<path::Path>,
+{
+    /// The credential helper.
+    credential: Credential,
+    /// The project URN.
+    urn: coco::Urn,
+    /// The branch to use for the checkout.
+    branch: String,
+    /// The path on the filesystem where we're going to checkout to.
+    path: P,
+    /// The `PATH` environment variable to be used for the checkout. It is safe to leave this
+    /// `None` when executing the application for real. However, if we want to run an integration
+    /// test we need to tell say where the `git-rad-remote` helper can be found.
+    bin_path: Option<ffi::OsString>,
+}
+
+impl<P> Checkout<P>
+where
+    P: AsRef<path::Path>,
+{
+    /// Create a new `Checkout` with the mock `Credential::Password` helper.
+    pub fn new<Bin>(urn: coco::Urn, branch: String, path: P, bin_path: Bin) -> Self
+    where
+        Bin: Into<Option<ffi::OsString>>,
+    {
+        Self {
+            // TODO(rudolfs): we'll have to figure out how to pass the secret
+            // key to git in a safe manner. As it is now it could be sniffed
+            // out from the process list while the user is doing a clone.
+            //
+            // How will we get ahold on the secret key here?
+            credential: Credential::Password("radicle-upstream".to_owned()),
+            urn,
+            branch,
+            path,
+            bin_path: bin_path.into(),
+        }
+    }
+
+    /// Checkout a working copy of a [`Project`].
+    ///
+    /// NOTE: `RAD_HOME` should be expected to be set if using a custom root for
+    /// [`librad::paths::Paths`]. If it is not set the underlying binary will delegate to the
+    /// `ProjectDirs` setup of the `Paths`.
+    ///
+    /// # Errors
+    ///
+    ///   * We couldn't resolve the executable path.
+    ///   * The checkout process failed.
+    pub fn run(self) -> Result<(), error::Error> {
+        let bin_path = match self.bin_path {
+            Some(path) => Ok(path),
+            None => Self::default_bin_path(),
+        }?;
+
+        let mut child_process = Command::new("git")
+            .arg("-c")
+            .arg(self.credential.to_helper())
+            .arg("clone")
+            .arg("-b")
+            .arg(self.branch)
+            .arg(LocalUrl::from(self.urn).to_string())
+            .arg(&self.path.as_ref().as_os_str())
+            .env("PATH", &bin_path)
+            .envs(std::env::vars().filter(|(key, _)| key.starts_with("GIT_TRACE")))
+            .spawn()?;
+
+        // TODO: Capture the error if any and respond
+        let result = child_process.wait()?;
+
+        if result.success() {
+            Ok(())
```

Comment:
Upon successfully cloning the repo to our directory we should set up the `remote = "rad"` which handles fetching and pushing from/to the monorepo. See `setup_remote` in `coco/peer.rs`

Reply:
The remote helper already creates a remote for us that points to the monorepo, it just called it by default `origin`. Renamed it to `rad` in: https://github.com/radicle-dev/radicle-upstream/pull/722/commits/8651198e24cdaa89aa70a370373766be42bd9274.